### ISSUE TITLE
Certificate expiry, API prior to 21

### ISSUE THREAD

**Issue by shahrukh802** [martykan/forecastie#487]
on below 21 devices the app opens and show the message network error please explain why it happens

**Comment by robinpaulson** [comment]
why do you think api 21 is the key problem?

**Comment by shahrukh802** [comment]
dont know it shows the message connection not available but it works fine on higher api devices

**Comment by robinpaulson** [comment]
hmm, it should work on api 15, as per the gradle build file. our error messages aren't great (see #316), so that message is probably misleading. which version of android is it running, precisely?

**Comment by FridoDeluxe** [comment]
@shahrukh802 Unfortunately the last commits which improve and clarify error messages are not released yet. To identify the problem you could
1. record logs when the error occurs and post them here
2. build the app yourself from the current source code and get better/more precise error messages
3. wait for the next release and try again (at least with more helpful error messages)

This generic network error message you are receiving right now could also be caused by an invalid OpenWeatherMap API key or a weather response that Forecastie is not able to decode properly.

**Comment by robinpaulson** [comment]
> Unfortunately the last commits which improve and clarify error messages are not released yet.

yep, point taken. new release out today.

**Comment by shahrukh802** [comment]
Thanks there is another issue the weather Map not moves with one finger on some devices

**Comment by FridoDeluxe** [comment]
> yep, point taken. new release out today.

It wasn't meant as a hidden reminder. I don't even know if it will help to solve this problem ...

**Comment by robinpaulson** [comment]
Oh, I realise it wasn't intended as a reminder, but I'd been a bit slack anyway, so figured I'd do it ASAP.

**Comment by robinpaulson** [comment]
> Thanks there is another issue the weather Map not moves with one finger on some devices

Could you be a bit more precise please, give a but more detail?

**Comment by FridoDeluxe** [comment]
@shahrukh802 [Forecastie 1.15.1 is available](https://f-droid.org/en/packages/cz.martykan.forecastie/).

**Comment by robinpaulson** [comment]
@shahrukh802 Is the original issue fixed now, with the new release?

**Comment by telochiasma** [comment]
Hello, I had the same problem on my rooted Note 2 (Kitkat 4.4.2) and found a solution:

tl;dr:
manually add current openweathermap.org certificates to your credential storage, as their server stopped supporting old ones on your device...

long story:
At the same time forcastie stopped working, because of before mentioned error, also some podcasts from podigee.io stopped downloading in antennapod. It gave the following error:

`IO Error: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found`

I figured I could maybe catch both of those flies at once. 
Here is my botched solution:

- You'll need a rooted device. 

- Get CAdroid (https://f-droid.org/en/packages/at.bitfire.cadroid/)

- with CAdroid download the top most certificate for domain https://openweathermap.org
 (for the new certificate to be downloaded you'll need to set up a screen lock: PIN or Password)

following these posts : 
https://android.stackexchange.com/a/111058
https://forum.xda-developers.com/showpost.php?p=48276260&postcount=6
... i got the following method:

- adb shell to your device (in usb debugging)
`$ su`
`# mount -o rw,remount /system`
`# cp /mnt/ext_card/data/misc/keychain/certs-added/<your new certificate> /system/etc/security/cacerts/`
`# cd /system/etc/security/cacerts/`
`# chmod 644 <your new certificate>`
`# reboot`

- reset to your previously used screen lock method

Forcastie should now be updating again.

**Comment by fAntel** [comment]
Thank you for pointing out the exact exception.

I think documentation for Android developers already has solution for this case: https://developer.android.com/training/articles/security-ssl.html#UnknownCa
and with next command
`echo -n | openssl s_client -connect api.openweathermap.org:443 | sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > openweathermap.crt`
we can get certificate file for OpenWeatherMap.

**Comment by robinpaulson** [comment]
Golly, thanks for figuring that out @telochiasma and @fAntel . Is there anyway we can incorporate any of those workarounds into Forecastie itself? Or perhaps we distribute the cert with Forecastie? I'm not at all up on certificates, so there could well be a giant hole in my cunning plan.

**Comment by fAntel** [comment]
I think solution that I've found need more researches to check compatibility with more recent Androids but it doesn't require rooted device. I think for now it is safe to use this solution only as response to this exception with repetition the request caused exception.

There is no source for CAdroid at f-droid page. If we find them or some other example, we even can add downloading of certificate in the app. Downloading certificate is better than put it into apk because it can be downloaded when it's expires without updating Forecastie.

**Comment by robinpaulson** [comment]
> There is no source for CAdroid at f-droid page.

It's five years old, but ....:

https://f-droid.org/repo/at.bitfire.cadroid_8_src.tar.gz

Yes to downloading cert rather than including in the .apk

**Comment by robinpaulson** [comment]
@fAntel 
>  we can get certificate file for OpenWeatherMap.

is there a particular mechanism for importing certificates, or do they simply need to be placed in a particular location?

**Comment by fAntel** [comment]
There is a code sample on the page I linked to. We can read it content with `FileInputStream` and put into `Certificate` class from many locations. And certificate is a public key, so I think we can choose whatever location we want. I prefer [app directory](https://developer.android.com/reference/kotlin/android/content/Context?hl=en#getFilesDir()) not to download it every time and on non-rooted devices it can't be changed by another application.

I’m still thinking about something. If we put the certificate in the apk, it could expire. But if we download it every time, could the app be attacked with Main in the Middle attack? Like a user connect to Wi-Fi in café where someone spams with wrong DNS resolves and Forecastie download wrong certificate. What private information we send to OWM? Only location, right? Because location of public Wi-Fi could be already known to attacker it's not a leak I suppose. I write it to someone could check am I right and certificate downloading is secure for users.

**Comment by robinpaulson** [comment]
@fantel did you get anywhere with your research into this?

**Comment by fAntel** [comment]
Oh, I hadn't understood who would do it. Yesterday I fixed this problem but put certificate into assets and look at ca-droid source code. It looks simple and I hope to implement downloading of certificate for Forecastie during this week.

**Comment by robinpaulson** [comment]
@fAntel You had a contact who was knowledgeable in this area that you were going to ask. Your final comment from July 10th

**Comment by fAntel** [comment]
Oh, I see. I have written it wrong. I mean may be somebody who would read this issue and knows more about this topic would be able to correct me or say that I am right.
With my knowledge solution with downloading certificate looks fine.

**Comment by robinpaulson** [comment]
@shahrukh802 We have included code which should fix the problem you are having. It will be available on F-Droid within a couple of days. Please let us know the outcome.
